src/app/clusters/network-commissioning/network-commissioning.cpp

```diff
@@ -0,0 +1,253 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "network-commissioning.h"
+
+#include <cstring>
+
+#include <lib/support/CodeUtils.h>
+#include <lib/support/ReturnMacros.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/ConnectivityManager.h>
+#include <transport/NetworkProvisioning.h>
+
+#if defined(CHIP_DEVICE_LAYER_TARGET)
+#define DEVICENETWORKPROVISIONING_HEADER <platform/CHIP_DEVICE_LAYER_TARGET/DeviceNetworkProvisioningDelegateImpl.h>
+#include DEVICENETWORKPROVISIONING_HEADER
+#endif
+
+namespace chip {
+namespace app {
+namespace clusters {
+namespace NetworkCommissioning {
+
+constexpr uint8_t kMaxNetworkIDLen       = 32;
+constexpr uint8_t kMaxThreadDatasetLen   = 128;
+constexpr uint8_t kMaxWiFiSSIDLen        = 32;
+constexpr uint8_t kMaxWiFiCredentialsLen = 64;
+constexpr uint8_t kMaxNetworks           = 4;
+constexpr char kThreadNetworkName[]      = "tmp-thread-network";
+
+// TODO: The network info should be stored in attribute storage instead of memory.
+// TODO: We cannot get length of operational dataset here, should be fixed after we have byte string type.
+// TODO: Should encode response for command responses.
+enum class NetworkCommissioninError : uint8_t
+{
+    kSuccess                = 0,
+    kOutOfRange             = 1,
+    kBoundsExceeded         = 2,
+    kNetworkIDNotFound      = 3,
+    kDuplicateNetworkID     = 4,
+    kNetworkNotFound        = 5,
+    kRegulatoryError        = 6,
+    kAuthFailure            = 7,
+    kUnsupportedSecurity    = 8,
+    kOtherConnectionFailure = 9,
+    kIPV6Failed             = 10,
+    kIPBindFailed           = 11,
+    kLabel9                 = 12,
+    kLabel10                = 13,
+    kLabel11                = 14,
+    kLabel12                = 15,
+    kLabel13                = 16,
+    kLabel14                = 17,
+    kLabel15                = 18,
+    kUnknownError           = 19,
+};
+
+enum class NetworkType : uint8_t
+{
+    kUndefined = 0,
+    kWiFi      = 1,
+    kThread    = 2,
+    kEthernet  = 3,
+};
+
+enum class WiFiSecurityType : uint8_t
+{
+    kUnencrypted  = 0,
+    kWEPPersonal  = 1,
+    kWPAPersonal  = 2,
+    kWPA2Personal = 3,
+    kWPA3Personal = 4,
+};
+
+struct ThreadNetworkInfo
+{
+    uint8_t mDataset[kMaxThreadDatasetLen];
+    uint8_t mDatasetLen;
+};
+
+struct WiFiNetworkInfo
+{
+    uint8_t mSSID[kMaxWiFiSSIDLen + 1];
+    uint8_t mSSIDLen;
+    uint8_t mCredentials[kMaxWiFiCredentialsLen];
+    uint8_t mCredentialsLen;
+    WiFiSecurityType mSecurity;
+};
+
+struct NetworkInfo
+{
+    uint8_t mNetworkID[kMaxNetworkIDLen];
+    uint8_t mNetworkIDLen;
+    uint8_t mEnabled;
+    NetworkType mNetworkType;
+    union NetworkData
+    {
+        ThreadNetworkInfo mThread;
+        WiFiNetworkInfo mWiFi;
+    } mData;
+};
+
+namespace {
+NetworkInfo sNetworks[kMaxNetworks];
```

Comment:
this will permanently use up some RAM - is this the intented final usage or are we supposed to migrate to some central 'connectivity' object to manage these?

We currently do not need this to join wifi at boot, so I imagine there should be a way too not need this.

Reply:
we supposed to migrate to some central 'connectivity' object to manage these. In fact, I'd like to move these to attribute storage if possible in the future updates.